Add hb_font_set_variations func

m4rc1e opened the pull request:
This pr adds the hb function [hb_font_set_variations](https://harfbuzz.github.io/harfbuzz-hb-font.html#hb-font-set-variations).

I've added a method to the font class so users can set variations using a dict in the following manner. 

```
font.variation({"wght": 300, "slnt": -20})
```

I'm unsure whether to keep the method name singular? the hb function is plural.

I need to implement this so I can generate diff images using freetype-py and this lib. If I need more variable related structs and functions, I'll add them. Also, thanks for making this binding. It's the best one imo.

anthrotype commented:
`set_variations` would be a better name, as it matches the one from hb

anthrotype commented on src/uharfbuzz/_harfbuzz.pyx:
```diff
@@ -287,6 +287,23 @@ cdef class Font:
+            variation.value = value
+            hb_variations[i] = variation
+        hb_font_set_variations(self._hb_font, hb_variations, size)
+        free(hb_variations)
```
maybe wrap the call to hb_font_set_variations in a `try: ... finally: free(hb_variations)` block.

anthrotype replied:
No, that’s not necessary. Actually, since we don’t expect the c++ function to raise an exception the try block is not needed here either, nevermind.

anthrotype replied:
on the other hand, the python statements in between the allocation of hb_variations and the call to free may raise exceptions, in which case we need to release it.
this is the standard pattern when using malloc/calloc/etc. from cython:
https://cython.readthedocs.io/en/latest/src/tutorial/memory_allocation.html
you'll notice that it also checks that malloc returns a NULL pointer in the unlikely event that you ran out of memory

anthrotype commented on src/uharfbuzz/_harfbuzz.pyx:
```diff
@@ -287,6 +287,23 @@ cdef class Font:
+        cdef unsigned int size
+        cdef hb_variation_t* hb_variations
+        cdef bytes packed
+        cdef char* cstr
```
this cstr variable seems unused

anthrotype commented:
Thanks Marc, looks good!

anthrotype commented:
I just updated the embedded harfbuzz library to the latest version (2.1.1) in the master branch, you may want to first rebase your PR branch on it to check everything still works.

m4rc1e commented:
Thanks for the review. I've squashed and repushed the following changes:

- method name changed to `set_variations`
- removed the unnecessary variable
- rebased with the latest hb
- implemented the try/finally malloc stuff.

For the last part, I just copied the example from the documentation. I have a vague idea of what's happening but you may want to look at this again.

anthrotype commented:
tagged 0.3.0 https://github.com/trufont/uharfbuzz/releases/tag/v0.3.0
wheels coming
